### PR TITLE
With custom HTML Renderer

### PR DESCRIPTION
**Pull request by kristojorg**
This is a draft PR showing what it would look like to remove R2D2BC.

TODO:

- [x] Color mode
- [x] Font size
- [x] Font family
- [x] Paginated mode
- [x] Snap to page in paginated mode
- [x] TOC navigation
- [x] Internal navigation links
- [x] Make internal link clicks register a change of location so prev/next page work
- [x] fix scroll bars
- [x] Remember where user is when switching between paginated and scroll mode
- [x] Scroll to end of resource when navigating backwards
- [x] Keep location in URL bar
- [x] paginated window resize
- [x] Fix bug when there is a "half page"

### CODE REVIEW

**Comment by kristojorg** [comment]
Pagination isn't working here: https://web-reader-git-refactor-remove-r2d2bc-nypl.vercel.app/v2/moby-epub2?location=%7B%22href%22%3A%22https%3A%2F%2Fweb-reader-git-refactor-remove-r2d2bc-nypl.vercel.app%2Fsamples%2Fmoby-epub2-exploded%2FOEBPS%2F%40public%40vhost%40g%40gutenberg%40html%40files%402701%402701-h%402701-h-1.htm.html%22%2C%22type%22%3A%22application%2Fxhtml%2Bxml%22%2C%22locations%22%3A%7B%22progression%22%3A0.7692144892144892%2C%22position%22%3A11%2C%22remainingPositions%22%3A2%7D%7D

**Comment by kristojorg** [comment]
@Toxiapo @hokei @samanthaandrews this PR is now ready for review. There is an architecture document at /HtmlReader/architecture.md. Take a look and please let me know your questions so I can update the arch document with answers!

**Comment by kristojorg** [comment]
I think once we have reviewed this, we should come up with a list of edge cases to add to our test suite so we can make sure this is running 100%

Top of my head:

- Fixed layout
- Initial location
- Reads location from query param on initial load. Handles multiple types of locators.
- Handles incoherent query param
- Internal link clicks (and location and query param update)
- Renders iframe title
- Forward/back buttons on browser work as expected (what is expected)
- hash navigation

File: src/HtmlReader/architecture.md
Comment (hokei) [inline src/HtmlReader/architecture.md:17]: Context:  So I'm starting my review/exploration with this file.  

Is it convention to call everything ThingState rather than ActiveState.FetchingResource (eg: https://github.com/NYPL-Simplified/web-reader/pull/91/files#diff-40c72fa84cd3e89b4e8825c1858a53ddb9d56a42805fcf97192b24aabbefe068R20)?

If someone were to say "oh my state is special, it's gotta be a new state" - where would they add it? How would they determine if that's actually true or not?

File: src/HtmlReader/architecture.md
Comment (hokei) [inline src/HtmlReader/architecture.md:27]: should those values stay strings or should we enum them

File: src/HtmlReader/architecture.md
Comment (hokei) [inline src/HtmlReader/architecture.md:42]: Is CSS the only injectible?  What other injectibles should we consider? Do these have implications for the rest of the update-resource process?

File: src/HtmlReader/architecture.md
Comment (hokei) [inline src/HtmlReader/architecture.md:46]: > we simple add it to the document string

nit: typo, but also:  what does the "it" here refer to?  The Readium-css stylesheet?

File: src/HtmlReader/architecture.md
Comment (kristojorg) [reply src/HtmlReader/architecture.md:27]: Enums aren't very nice in TS unfortunately, and strings provide all the same benefits. They are _typed_ strings though, so you can't accidentally put something different in there

File: src/HtmlReader/architecture.md
Comment (kristojorg) [reply src/HtmlReader/architecture.md:17]: That's a good question, I should think about this a little more. You're asking whether these are all a "sub-state" of the ActiveState?

In terms of how to know when you need a new one... It's a little bit arbitrary. When I first built this, I didn't have these split into multiple states. It was all just one `ReaderState`, but then I constantly had to check `if(state.isIframeLoaded && state.iframe && state.resource)` etc. So to minimize the number of those checks, I separated them into distinct states. Then we know that `if(state.state === 'READY')` then all those things are defined. It's just a convenient way to separate the complex number of options into distinct states so we can know what is defined at different times.

File: src/HtmlReader/architecture.md
Comment (kristojorg) [reply src/HtmlReader/architecture.md:42]: No, another injectable is JS or, theoretically, any element that can go in the head of the document. Typically, we should be prepared for CSS and JS injections. They don't have implications because we don't support _changing_ the injectables after initialization. We don't need that, and I think if someone wanted it, we would let them do it with some kind of API. But for now, no, the injectables are injected only when the resource is fetched.

File: src/HtmlReader/architecture.md
Comment (kristojorg) [reply src/HtmlReader/architecture.md:46]: Yea, I meant "we simply add all injectables to the document before it gets to the iframe".

Comment (hokei) [reply src/HtmlReader/architecture.md:46]: OK so it goes document -> put injectibles into document -> document + injectibles get to the iFrame.  And that's we make sure things are styled on first render.  Got it.

File: src/HtmlReader/architecture.md
Comment (hokei) [reply src/HtmlReader/architecture.md:17]: > You're asking whether these are all a "sub-state" of the ActiveState?

Yes, or if they're all a "sub-state" of _something_.  clearly InactiveState isn't a substate of ActiveState.  The idea is to try to put all this together so that it's referenced everywhere, so that, for example, we don't have a chance of doing something like `state.state === 'NOT_QUITE_READY'` in a random place in the code.  I haven't looked through everything in detail so you might have put guardrails against that already.

Comment (hokei) [reply src/HtmlReader/architecture.md:17]: In essence, understanding all the states the reader can be in is the most important thing, so I'm wondering if there's a way we can define all of it in one place in the code, as well as in the documentation.  ActiveState seemed like the most obvious place to do it.

File: src/HtmlReader/lib.ts
Comment (hokei) [inline src/HtmlReader/lib.ts:17]: what is the difference between fixed injectibles and non-fixed injectibles?

File: src/HtmlReader/index.tsx
Comment (hokei) [inline src/HtmlReader/index.tsx]: where are these importing from?  I don't think I've ever seen this before haha

File: src/HtmlReader/index.tsx
Comment (hokei) [inline src/HtmlReader/index.tsx:35]: Why?

File: src/HtmlReader/index.tsx
Comment (hokei) [inline src/HtmlReader/index.tsx:33]: Are these all things that work in the previous reader?  If not, worth thinking about extracting them and letting POs prioritize

File: src/HtmlReader/index.tsx
Comment (kristojorg) [reply src/HtmlReader/index.tsx:35]: It doesn't support fixed layout. Instead you rely on the publisher's CSS

File: src/HtmlReader/index.tsx
Comment (kristojorg) [reply src/HtmlReader/index.tsx]: Should be the index two levels up I guess? I'm not sure why they auto imported from there. They are also defined in constants

File: src/HtmlReader/lib.ts
Comment (kristojorg) [reply src/HtmlReader/lib.ts:17]: One is for fixed layout and the other is for reflowable books

File: src/HtmlReader/architecture.md
Comment (kristojorg) [reply src/HtmlReader/architecture.md:46]: Exactly

File: src/HtmlReader/architecture.md
Comment (kristojorg) [reply src/HtmlReader/architecture.md:17]: I suppose that's what I was trying to do in the `/HtmlReader/types.ts` file where I defined all the state types. Do you think that works?

File: src/HtmlReader/index.tsx
Comment (kristojorg) [reply src/HtmlReader/index.tsx:33]: No, I'm pretty sure this PR is now feature-paired with prod. We should probably do some fixed layout testing though come to think of it. I need to update these comments

File: src/HtmlReader/architecture.md
Comment (Toxiapo) [inline src/HtmlReader/architecture.md:63]: What does CFI stand for?

File: src/HtmlReader/architecture.md
Comment (Toxiapo) [inline src/HtmlReader/architecture.md:75]: This is great that we made this explicit! This makes things a lot easier (i.e. when comparing internal links. 😆 )

File: src/HtmlReader/effects.ts
Comment (Toxiapo) [inline src/HtmlReader/effects.ts:99]: 1 column FTW 🎉

File: src/HtmlReader/lib.ts
Comment (Toxiapo) [inline src/HtmlReader/lib.ts]: nit: dangling comment

File: src/HtmlReader/architecture.md
Comment (kristojorg) [reply src/HtmlReader/architecture.md:63]: Canonical Fragment Identifier. I'm not totally sure how they work, but basically it's some string that points to a specific position in an HTML document.

File: src/HtmlReader/architecture.md
Comment (kristojorg) [reply src/HtmlReader/architecture.md:75]: Yea that was a pain previously!

File: src/HtmlReader/lib.ts
Comment (hokei) [inline src/HtmlReader/lib.ts:232]: dumb question - there truly isn't a better way to do an optional?

Comment (hokei) [reply src/HtmlReader/lib.ts:232]: Is this so that you don't have to check for `undefined` every time you use `getIframeHtml`?

**Review by hokei — APPROVED** [review]
This looks really good.  

Do we have a good set of acceptance criteria on how to test this, to make sure that it truly is feature-parity with the previous reader?  

It can be manual tests.  

(This can be a new ticket) 
Do we have automated test suites for all of our locations logic?  I see that navigations has page button stuff, but (it might be there but I might be missing it?) something that tests things like, when you scroll to the bottom of the chapter and press "next", it knows that it's at the end of the chapter and loads the next one.  Or, when you scroll, then resize, it knows you're not yet at the end of the chapter.  

(Or are these things no longer edge cases?) 

Also, do we need a tests for the example app, for things like making sure that embedded mode continues to work?

**Comment by hokei** [comment]
(not) fun scrollbar bug, I think
Screen Shot 2022-01-24 at 1 45 17 PM

  (this is chrome with a mouse plugged in - you can't see it in regular view because it's white) 
  
 (This is the kind of thing that I don't really know how to test for, but also is the kind of thing I worry about regressions on)

File: src/HtmlReader/useResource.ts
Comment (samanthaandrews) [inline src/HtmlReader/useResource.ts]: nit: remove unused injectJS() since we are calling it above

**Comment by Toxiapo** [comment]
> This looks really good.
> 
> Do we have a good set of acceptance criteria on how to test this, to make sure that it truly is feature-parity with the previous reader?
> 
> It can be manual tests.
> 
> (This can be a new ticket) Do we have automated test suites for all of our locations logic? I see that navigations has page button stuff, but (it might be there but I might be missing it?) something that tests things like, when you scroll to the bottom of the chapter and press "next", it knows that it's at the end of the chapter and loads the next one. Or, when you scroll, then resize, it knows you're not yet at the end of the chapter.
> 
> (Or are these things no longer edge cases?)
> 
> Also, do we need a tests for the example app, for things like making sure that embedded mode continues to work?

Yeah, I agree! I think some manual tests are good for this PR and getting more tests coverages in the future.

(Thinking out loud on some of the issues we had faced previously)

1. Internal links works on both paginated and scrolling mode
2. TOC navigates on paginated and scrolling mode
3. Height and width problem, make sure there are no extra scrollbars 
4. All the settings under the settings button
5. Next / Previous button response to book state (navigates + hide/show state), i.e stuck on cover image issue
6. Switch from paginated to scrolling mode saves the reading position, vice versa

File: src/HtmlReader/lib.ts
Comment (kristojorg) [reply src/HtmlReader/lib.ts:232]: Exactly. I wanted one function for when it _has_ to be defined, and another for when maybe it's defined and maybe not.

**Comment by kristojorg** [comment]
> Do we have a good set of acceptance criteria on how to test this, to make sure that it truly is feature-parity with the previous reader?

I don't have anything concrete, but Jiayong's list is a good starting point. I would just add fixed layout to that. Is this something I should do or I should have the QA team do?

I updated [OE-321](https://jira.nypl.org/browse/OE-321) to add automated tests for these edge cases you listed above. I think that will be very helpful. Probably won't happen for me until next sprint though.

**Comment by kristojorg** [comment]
@hokei RE that scroll bar bug:

Soooo I don't think this is a regression. Question: Can you actually scroll horizontally? How much if so? The issue is that when scroll bars are showing, which you see when you have a mouse plugged in (can also enable it in system preferences/general), the vertical scroll bar actually makes the available space to render smaller, but is not included in `window.innerWidth` or `clientWidth`. So when we calculate the width of the window to render, we might be off by a certain amount due to the vertical scrollbar. Which causes a horizontal scrollbar to be rendered (lol). We _could_ try to figure out if it is showing, but then there is the secondary issue that it is actually a different size on each platform/browser combo. Sooooo I just decided it's not worth it basically. If you're using the app with scrollbars showing like that, you're probably used to seeing scrollbars everywhere even when there is very little area to scroll. It seems like a platform problem to me

**Comment by Toxiapo** [comment]
I am seeing a UI bug with the internal link, this does not affect the user's ability to read the book, b/c it's fine if you refresh the page or change the page using next/prev button, so maybe we can fix this after the merge? Unless this is a quick fix.

To reproduce:
1. Go on /html/moby-epub2
2. Go to the page with internal links, and click on one
3. The page navigates but the page layout is misaligned.  

Screen Shot 2022-01-26 at 10 02 50 AM

**Comment by hokei** [comment]
>  Question: Can you actually scroll horizontally? How much if so? 

Screen Shot 2022-01-26 at 10 11 28 AM

So yeah it's a bit of a strange bug in that I see the scrollbars area but I cannot scroll.  I've showed it in the embedded view which makes this super clear that scrolling isn't actually possible. I also think it isn't a high priority bug and should not block merge.  Honestly I think very few people will notice it and we can fix it when they start filing bug reports (or when I get annoyed enough with it haha)

**Comment by hokei** [comment]
I tried putting this in next apps (CPW and DRB) and got this error: 
Screen Shot 2022-01-26 at 3 46 37 PM

Does there need to be updates to the Readme on how to incorporate the CSS?

File: src/HtmlReader/index.tsx
Comment (kristojorg) [reply src/HtmlReader/index.tsx]: Fixed this

File: cypress/integration/html-reader/navigations.ts
Comment (Toxiapo) [inline cypress/integration/html-reader/navigations.ts]: Do we still need to do this?

File: .github/workflows/cypress.yml
Comment (Toxiapo) [inline .github/workflows/cypress.yml]: I am curious what this does? How does the CI install other packages?

File: cypress/integration/html-reader/errors.ts
Comment (Toxiapo) [inline cypress/integration/html-reader/errors.ts]: We probably don't need this since we will always have default injectables right?

File: src/Readium/Locator.ts
Comment (samanthaandrews) [inline src/Readium/Locator.ts]: nit: remove displayInfo

**Comment by hokei** [comment]
Has anyone been able to `npm link` this into a consuming package such as CPW?  I just want to make sure that this has been tested - I think I'm running into react version errors, which might be  npm link errors rather than errors in the new code, so I don't want to that to hold up merge -- just let me know

File: cypress/integration/html-reader/navigations.ts
Comment (kristojorg) [reply cypress/integration/html-reader/navigations.ts]: Good point

**Comment by kristojorg** [comment]
>Has anyone been able to npm link this into a consuming package such as CPW? I just want to make sure that this has been tested - I think I'm running into react version errors, which might be npm link errors rather than errors in the new code, so I don't want to that to hold up merge -- just let me know

No I haven't either. It's notoriously difficult to npm link with react. Involves a lot of BS. I am planning to publish an alpha and test with that via npm in CPW

**Comment by hokei** [comment]
> Involves a lot of BS. I am planning to publish an alpha and test with that via npm in CPW

OK that sounds like a good plan! Let me know and I'll test it on DRB too.

File: src/types.ts
Comment (Toxiapo) [inline src/types.ts]: old d-i-t-a reference

**Review by samanthaandrews — APPROVED** [review]
I have done a bunch of manual testing as well as reviewing the code and this looks really good. I know @Toxiapo and @hokei have more experience with this code base, and this PR has gone through thorough code review by them.

**Comment by kristojorg** [comment]
@hokei I have published as `@nypl/web-reader@1.1.0-0`. Install that version to test. I tested in CPW and all is working well so far.

**Comment by hokei** [comment]
This looks really good, and I think we should merge it.  However, I'll note down the things that I noticed.  I don't think we need to maintain two forks or anything - just that between OE and DRB we need to figure out when to fix these things before DRB can upgrade.  

Scrollbars:  left is current reader, right is new reader.  
Screen Shot 2022-03-07 at 9 33 14 AM

TOC: 
Screen Shot 2022-03-07 at 9 34 00 AM
All of the TOC items seem to have two borders instead of one ( this is a problem with everything, but it's most noticeable in the TOC) 

And then, 1 more that I think is an issue in both the old and new renderers, but you can see it in the new renderer in the screenshot above: the TOC should be above the prev/next buttons, but they're not. 


aaaaand that's really all I can find!  This is amazingly impressive work and let's get it in!

**Comment by hokei** [comment]
Also Kristo - after you merge, can you write up a little thing that's just what this feature will do for us?  Something that we can share with Miguel and Sean - what is now possible that was impossible before?  Especially for Sean, who is probably more likely to be making WR changes in the next 6-12 months

**Review by Toxiapo — APPROVED** [review]

**Review by hokei — APPROVED** [review]

**Pull request merged**
